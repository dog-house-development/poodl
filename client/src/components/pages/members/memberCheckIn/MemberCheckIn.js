import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import _ from 'lodash';
import classnames from 'classnames';

import MemberActions from '../../../../actions/memberActions';
// import ActivityActions from '../../../../actions/activityActions';
// import Loading from '../../../ui/Loading';
import Button from '../../../ui/Button';

// check-in pages
import FindMember from './pages/FindMember';
import SelectActivities from './pages/SelectActivities';
import Finished from './pages/Finished';

const pages = {
    findMember: { title: 'FindMember', index: 0 },
    selectActivities: { title: 'SelectActivities', index: 1 },
    finished: { title: 'Finished', index: 2 }
};

class MemberCheckIn extends Component {
    constructor(props) {
        super(props);
        this.state = {
            currentPage: pages.findMember,
            memberId: null
        };
        this.onNextClick = this.onNextClick.bind(this);
        this.setMemberId = this.setMemberId.bind(this);
    }

    setMemberId(id) {
        this.setState({ memberId: id, currentPage: pages.selectActivities });
    }

    getBackButtonMarkup() {
        if (this.state.currentPage !== pages.findMember) {
            return (
                <Button
                    onClick={() =>
                        this.setState({ currentPage: _.find(pages, { index: this.state.currentPage.index - 1 }) })
                    }
                    kind="secondary">
                    <i className="material-icons">arrow_back</i> Back
                </Button>
            );
        }
    }

    onNextClick() {
        if (this.state.currentPage === pages.finished) {
            this.setState({ currentPage: pages.findMember });
        } else {
            this.setState({ currentPage: _.find(pages, { index: this.state.currentPage.index + 1 }) });
        }
        // update the member checkins
        if (this.state.currentPage === pages.selectActivities) {
            this.props.memberActions.edit(this.state.memberId, {
                checkIns: [..._.get(this.props.member, 'checkIns', []), new Date()]
            });
        }
    }

    getNextButtonMarkup() {
        if (this.state.currentPage !== pages.findMember) {
            return (
                <Button onClick={this.onNextClick} kind="primary">
                    {this.state.currentPage === pages.finished ? (
                        <>
                            Start over <i className="material-icons">subdirectory_arrow_left</i>
                        </>
                    ) : (
                        <>
                            {this.state.currentPage === pages.selectActivities ? 'Finish' : 'Next'}{' '}
                            <i className="material-icons">arrow_forward</i>
                        </>
                    )}
                </Button>
            );
        }
    }

    getPageMarkup() {
        switch (this.state.currentPage) {
            case pages.findMember:
                return (
                    <div>
                        <FindMember setMemberId={this.setMemberId} />
                    </div>
                );
            case pages.selectActivities:
                return (
                    <div>
                        <SelectActivities memberId={this.state.memberId} />
                    </div>
                );
            case pages.finished:
                return (
                    <div>
                        <Finished />
                    </div>
                );
            default:
                return <p>you are lost my friend</p>;
        }
    }

    render() {
        return (
            <div className="member-check-in-container">
                <div>
                    <h1>Check in</h1>
                    <div
                        className={classnames('check-in-panel', {
                            panel: this.state.currentPage !== pages.selectActivities
                        })}>
                        {this.getPageMarkup()}
                    </div>
                    <div className="navigation-buttons-container">
                        <div className="back-button-container">{this.getBackButtonMarkup()}</div>
                        <div className="next-button-container">{this.getNextButtonMarkup()}</div>
                    </div>
                </div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        member: state.members.all[props.match.params.id],
        loading: state.members.loading,
        errors: state.members.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        memberActions: bindActionCreators(MemberActions, dispatch)
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MemberCheckIn);

// export default MemberCheckIn;
