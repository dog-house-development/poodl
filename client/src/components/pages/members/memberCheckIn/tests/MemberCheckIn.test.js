import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';
import { MemberCheckIn, mapStateToProps, mapDispatchToProps, pages } from '../MemberCheckIn';

configure({ adapter: new Adapter() });

describe('MemberCheckIn tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                members: {
                    all: {
                        123: {
                            _id: '123',
                            firstName: 'Big',
                            lastName: 'Tup',
                            membershipDate: '02/20/2020',
                            email: 'bigtup@nowhere.com'
                        }
                    },
                    errors: {},
                    loading: false
                }
            },
            _state
        );
        props = _.assign({}, { history: { listen: _.noop } }, _props);
        wrapper = shallow(
            <MemberCheckIn
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
        instance.state = { memberId: null, currentPage: pages.findMember };
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                members: {
                    123: {
                        _id: '123',
                        firstName: 'Big',
                        lastName: 'Tup',
                        membershipDate: '02/20/2020',
                        email: 'bigtup@nowhere.com'
                    }
                },
                loading: false,
                errors: {}
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(
                JSON.stringify({ memberActions: {}, authActions: {} })
            );
        });
    });

    describe('setMemberId', () => {
        it('should set member id and current page to select activities', () => {
            instance.setMemberId('66');
            expect(instance.state).toEqual({ currentPage: pages.selectActivities, memberId: '66' });
        });
    });

    describe('getBackButtonMarkup', () => {
        it('should not return markup if current page is find member or member sign up', () => {
            instance.setState({ currentPage: pages.findMember });
            expect(instance.getBackButtonMarkup()).toBeUndefined();
            instance.setState({ currentPage: pages.memberSignUp });
            expect(instance.getBackButtonMarkup()).toBeUndefined();
        });

        it('should return back button markup', () => {
            instance.setState({ currentPage: pages.selectActivities });
            expect(instance.getBackButtonMarkup()).toMatchSnapshot();
        });
    });

    describe('onNextClick', () => {
        it('should go to find member page if on finished page', () => {
            instance.setState({ currentPage: pages.finished });
            instance.onNextClick();
            expect(instance.state.currentPage).toEqual(pages.findMember);
        });

        it('should go to next page if not on finished page', () => {
            instance.setState({ currentPage: pages.selectActivities });
            instance.onNextClick();
            expect(instance.state.currentPage).toEqual(pages.finished);
        });
    });

    describe('getNextButtonMarkup', () => {
        it('should not return markup if current page is find member or member sign up', () => {
            instance.setState({ currentPage: pages.findMember });
            expect(instance.getNextButtonMarkup()).toBeUndefined();
            instance.setState({ currentPage: pages.memberSignUp });
            expect(instance.getNextButtonMarkup()).toBeUndefined();
        });

        it('should return next button markup', () => {
            instance.setState({ currentPage: pages.selectActivities });
            expect(instance.getNextButtonMarkup()).toMatchSnapshot();
        });
    });

    describe('onSignUpClick', () => {
        it('should set current page to member sign up', () => {
            instance.onSignUpClick();
            expect(instance.state).toEqual({ currentPage: pages.memberSignUp, memberId: null });
        });
    });

    describe('onSignUpSuccess', () => {
        it('should set memberId and change page to find member', () => {
            instance.onSignUpSuccess('123');
            expect(instance.state).toEqual({ memberId: '123', currentPage: pages.findMember });
        });
    });

    describe('getPageMarkup', () => {
        it('should return memberSignUpMarkup', () => {
            instance.setState({ currentPage: pages.memberSignUp });
            expect(instance.getPageMarkup()).toMatchSnapshot();
        });

        it('should return findMember', () => {
            instance.setState({ currentPage: pages.findMember });
            expect(instance.getPageMarkup()).toMatchSnapshot();
        });

        it('should return selectActivities', () => {
            instance.setState({ currentPage: pages.selectActivities });
            expect(instance.getPageMarkup()).toMatchSnapshot();
        });

        it('should return finished', () => {
            instance.setState({ currentPage: pages.finished });
            expect(instance.getPageMarkup()).toMatchSnapshot();
        });

        it('should return you are lost', () => {
            instance.setState({ currentPage: {} });
            expect(instance.getPageMarkup()).toMatchSnapshot();
        });
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
