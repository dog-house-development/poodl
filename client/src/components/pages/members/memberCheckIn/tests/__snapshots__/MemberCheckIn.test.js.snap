// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberCheckIn tests getBackButtonMarkup should return back button markup 1`] = `
<Button
  kind="secondary"
  onClick={[Function]}
  size="medium"
  type="button"
>
  <i
    className="material-icons"
  >
    arrow_back
  </i>
   Back
</Button>
`;

exports[`MemberCheckIn tests getNextButtonMarkup should return next button markup 1`] = `
<Button
  kind="primary"
  onClick={[Function]}
  size="medium"
  type="button"
>
  <React.Fragment>
    Finish
     
    <i
      className="material-icons"
    >
      arrow_forward
    </i>
  </React.Fragment>
</Button>
`;

exports[`MemberCheckIn tests getPageMarkup should return findMember 1`] = `
<div>
  <Connect(FindMember)
    onSignUpClick={[Function]}
    setMemberId={[Function]}
  />
</div>
`;

exports[`MemberCheckIn tests getPageMarkup should return finished 1`] = `
<div>
  <Finished />
</div>
`;

exports[`MemberCheckIn tests getPageMarkup should return memberSignUpMarkup 1`] = `
<div>
  <Connect(SelfRegisterMember)
    onSignUpSuccess={[Function]}
  />
</div>
`;

exports[`MemberCheckIn tests getPageMarkup should return selectActivities 1`] = `
<div>
  <Connect(SelectActivities)
    memberId={null}
  />
</div>
`;

exports[`MemberCheckIn tests getPageMarkup should return you are lost 1`] = `
<p>
  you are lost my friend
</p>
`;

exports[`MemberCheckIn tests render should render correctly 1`] = `
<div
  className="member-check-in-container"
>
  <div>
    <div
      className="check-in-panel"
    >
      <div>
        <Connect(FindMember)
          onSignUpClick={[Function]}
          setMemberId={[Function]}
        />
      </div>
    </div>
    <div
      className="navigation-buttons-container"
    >
      <div
        className="back-button-container"
      />
      <div
        className="next-button-container"
      />
    </div>
  </div>
</div>
`;
